source/nijigenerate/io/inimport.d

```diff
@@ -0,0 +1,267 @@
+        }
+
+    }
+
+    return outLayers;
+}
+
+/**
+    Imports a PSD file with user prompt.
+    also see incAskImportKRA()
+*/
+bool incAskImport(T)(string file) {
+    if (!file) return false;
+
+    auto handler = new LoadHandler!T(file);
+    return incKeepStructDialog(handler);
+}
+
+class LoadHandler(T) : ImportKeepHandler {
+    private string file;
+
+    this(string file) {
+        super();
+        this.file = file;
+    }
+
+    override
+    bool load(AskKeepLayerFolder select) {
+        switch (select) {
+            case AskKeepLayerFolder.NotPreserve:
+                incImport!T(file, IncImportSettings(false));
+                return true;
+            case AskKeepLayerFolder.Preserve:
+                incImport!T(file, IncImportSettings(true));
+                return true;
+            case AskKeepLayerFolder.Cancel:
+                return false;
+            default:
+                throw new Exception("Invalid selection");
+        }
+    }
+}
+
+/**
+    Imports a PSD file.
+    Note: You should invoke incAskImportPSD for UI interaction.
+*/
+void incImport(T)(string file, IncImportSettings settings = IncImportSettings.init) {
+    incNewProject();
+    // TODO: Split this up to a seperate file and make it cleaner
+    try {
+
+        T doc = Traits!T.parseDocument(file);
+        IncImportLayer!T[] layers = incBuildLayerLayout!T(doc);
+        vec2i docCenter = vec2i(doc.width/2, doc.height/2);
+        Puppet puppet = new ExPuppet();
+
+        void recurseAdd(Node n, IncImportLayer!T layer) {
+            
+            Node child;
+            if (layer.isLayerGroup) {
+                if (settings.keepStructure) {
+                    child = inInstantiateNode(settings.layerGroupNodeType, cast(Node)null);
+                    if (auto part = cast(Part)child) part.blendingMode = layer.blendMode;
+                }
+            } else {
+                
+                layer.imageLayerRef.extractLayerImage();
+                inTexPremultiply(layer.imageLayerRef.data);
+                auto tex = new Texture(layer.imageLayerRef.data, layer.imageLayerRef.width, layer.imageLayerRef.height);
+                ExPart part = incCreateExPart(tex, null, layer.name);
+                part.layerPath = layer.getLayerPath();
+
+                auto layerSize = cast(int[2])layer.imageLayerRef.size();
+                vec2i layerPosition = vec2i(
+                    layer.imageLayerRef.left,
+                    layer.imageLayerRef.top
+                );
+
+                // TODO: more intelligent placement
+                part.localTransform.translation = vec3(
+                    (layerPosition.x+(layerSize[0]/2))-docCenter.x,
+                    (layerPosition.y+(layerSize[1]/2))-docCenter.y,
+                    0
+                );
+
+
+                part.setEnabled(Traits!T.isVisible(layer.imageLayerRef));
+                part.opacity = (cast(float)layer.imageLayerRef.opacity)/255;
```

Comment:
I think this code could be refactored out of the `if-else` block to allow support for opacity for Groups.

Reply:
Currently, opacity is implemented at Part and Composite separately.
Condition check might be different from that of non-group once.
So I guess adding another opacity assignment is appropriate rather than moving the code.
e.g. 
```
if (auto part = cast(Part)child) { part.blendingMode = layer.blendMode; part.opacity = (...); }
else if (auto comp = cast(Composite)child) { comp.opacity = (...); }
```

Reply:
Yes, but I think that logic for `Part` both group and non-group might be the same. should be
```d
if (auto part = cast(Part)child) {
    part.setEnabled(Traits!T.isVisible(layer.imageLayerRef));
    part.opacity = (cast(float)layer.imageLayerRef.opacity)/255;
    part.blendingMode = layer.blendMode;
} else if (auto comp = cast(Composite)child) {
    ...
}
```
I think I might write it as a function. It should solve both opacity and disabling for groups, with potential for adding other code in the future.

In the future, we might consider using an interface or further abstracting it.